src/DevTo.Api/ListingsService.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.Linq;
+using System.Net.Http;
+using System.Reflection;
+using System.Threading.Tasks;
+using Forem.Api.Models;
+using Forem.Api.Models.Listing;
+using Newtonsoft.Json;
+
+namespace Forem.Api
+{
+	public class ListingsService : ApiService, IListingsApi
+	{
+		public ListingsService(Uri baseUri, HttpClient httpClient) : base(baseUri, httpClient) { }
+
+		public Task<IEnumerable<Listing>> GetListingsAsync(int page = 1, int perPage = 30)
+			=> GetAsync<IEnumerable<Listing>>("/api/listings", new Dictionary<string, object>
+				{
+					{ "page", page },
+					{ "per_page", perPage },
+				});
+
+		public Task<Listing> CreateListingAsync(string apiKey, ListingProperties properties, ListingAction action = ListingAction.Create)
+			=> PostAsync<Listing>("/api/listings", CreateRequestPayload(properties, action), apiKey);
+
+		public Task<IEnumerable<Listing>> GetListingsByCategoryAsync(ListingCategory category, int page = 1, int perPage = 30)
+			=> GetAsync<IEnumerable<Listing>>($"/api/listings/category/{category}", new Dictionary<string, object>
+				{
+					{ "page", page },
+					{ "per_page", perPage },
+				});
+
+		public Task<Listing> GetListingAsync(long id) => GetAsync<Listing>($"/api/listings/{id}");
+
+		public Task<CompleteListing> UpdateListingAsync(string apiKey, long id, ListingAction action, ListingProperties properties = null)
+			=> PutAsync<CompleteListing>($"/api/listings/{id}", CreateRequestPayload(properties, action), apiKey);
+
+		private dynamic CreateRequestPayload(ListingProperties properties, ListingAction action)
+		{
+			var listing = new ExpandoObject() as IDictionary<string, object>;
+
+			switch (action)
+			{
+				case ListingAction.Bump:
+				case ListingAction.Publish:
+				case ListingAction.Unpublish:
+					listing.Add("action", action);
+					break;
+
+				case ListingAction.Create:
+				case ListingAction.Draft:
+				case ListingAction.Update:
+					if (action == ListingAction.Draft)
+					{
+						listing.Add("action", action);
+					}
+
+					var props = properties.GetType().GetProperties();
```

Comment:
I'd probably stick away from doing reflection here. While it is less code than checking each property of `ListingProperties` manually, I feel it does a disservice for understanding what is going on. While it might seem we're then doubling up on the JSON property names that you've defined on the model, from the looks of it, we don't actually have the JSON serializer convert those types so their strings could just be referenced here instead.